mypy doesn't narrow `Final` objects in nested contexts

**Bug Report**

I have a global constant I've marked with `Final`. In a class definition, I use this constant to swap out implementations of a method. While inside a type guarding block, but outside the method definition, the type of the `Final` object is narrowed, but inside of the method it is not narrowed.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&gist=2d4c97cffea1dd6710bf5cc7283f7290

```python
from typing import *

def _init() -> None | str:
    return None

RESOLVE_X: Final = _init()

class Example:
    
    if RESOLVE_X is not None:
        
        reveal_type(RESOLVE_X)  # note: Revealed type is "builtins.str"
        
        def __str__(self) -> str:
            return RESOLVE_X  # error: Incompatible return value type (got "str | None", expected "str")  [return-value]
            
    else:
        def __str__(self) -> str:
            return "example"
```

**Expected Behavior**

Inside the `__str__` definition in the `if RESOLVE_X is not None:` block, RESOLVE_X should stay narrowed as it cannot be reassigned and have its type re-widened in the future.

**Actual Behavior**

```
main.py:12: note: Revealed type is "builtins.str"
main.py:15: error: Incompatible return value type (got "str | None", expected "str")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

Mypy version 1.15.0.
Python 3.12.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -1557,6 +1557,10 @@ class TypeChecker(NodeVisitor[None], TypeCheckerSharedApi):
         Note that this doesn't do a full CFG analysis but uses a line number based
         heuristic that isn't correct in some (rare) cases.
         """
+        if v.is_final:
+            # Final vars are definitely never reassigned.
+            return False
+
         outers = self.tscope.outer_functions()
         if not outers:
             # Top-level function -- outer context is top level, and we can't reason about

--- a/test-data/unit/check-final.test
+++ b/test-data/unit/check-final.test
@@ -1258,9 +1258,6 @@ def _init() -> Union[int, None]:
     return 0
 
 FOO: Final = _init()
-# For some reason the following isn't narrowed at all
-# (always stays int | None):
-# FOO: Final[Union[int, None]] = 0
 
 class Example:
 

--- a/test-data/unit/check-final.test
+++ b/test-data/unit/check-final.test
@@ -1250,3 +1250,30 @@ def check_final_init() -> None:
     new_instance = FinalInit()
     new_instance.__init__()
 [builtins fixtures/tuple.pyi]
+
+[case testNarrowingOfFinalPersistsInFunctions]
+from typing import Final, Union
+
+def _init() -> Union[int, None]:
+    return 0
+
+FOO: Final = _init()
+# For some reason the following isn't narrowed at all
+# (always stays int | None):
+# FOO: Final[Union[int, None]] = 0
+
+class Example:
+
+    if FOO is not None:
+        reveal_type(FOO)  # N: Revealed type is "builtins.int"
+
+        def fn(self) -> int:
+            return FOO
+
+if FOO is not None:
+    reveal_type(FOO)  # N: Revealed type is "builtins.int"
+
+    def func() -> int:
+        return FOO
+
+[builtins fixtures/tuple.pyi]

--- a/test-data/unit/check-final.test
+++ b/test-data/unit/check-final.test
@@ -1275,5 +1275,3 @@ if FOO is not None:
 
     def func() -> int:
         return FOO
-
-[builtins fixtures/tuple.pyi]

